src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -393,6 +407,84 @@ private static (T chosen, EntryState state) GetModifiedItemAndState(T previous,
+            => queue.TryDequeue(out var item) ? item : (ArrayBuilder<TValue>.GetInstance(), new PoolingStatistics());
+
+        private static void ReturnPooledItem<TValue>(
+            ConcurrentQueue<(ArrayBuilder<TValue> builder, PoolingStatistics statistics)> queue,
+            (ArrayBuilder<TValue> builder, PoolingStatistics statistics) item)
+        {
+            // Don't bother shrinking the array for arrays less than this capacity.  They're not going to be a
+            // huge waste of space so we can just pool them forever.
+            const int MinCapacityToConsiderThreshold = 1000;
+
+            // The number of times something is added/removed to the pool before we start considering
+            // statistics. This is so that we have enough data to reasonably see if something is consistently
+            // sparse.
+            const int MinTimesPooledToConsiderStatistics = 100;
+
+            // The ratio of Count/Capacity to be at to be considered sparse.  under this, there is a lot of
+            // wasted space and we would prefer to just throw the array away.  Above this and we're reasonably
+            // filling the array and should keep it around.
+            const double SparseThresholdRatio = 0.25;
+
+            // The ratio of times we pooled something sparse.  Once above this, we will jettison the array as
+            // being not worth keeping.
+            const double ConsistentlySparseRatio = 0.75;
+
+            // Note: the values 0.25 and 0.75 were picked as they reflect the common array practicing of growing
+            // by doubling.  So once we've grown so much that we're consistently under 25% of the array, then we
+            // want to shrink down.  To prevent shrinking and inflating over and over again, we only shrink when
+            // we're highly confident we're going to stay small.
+
+            var (builder, statistics) = item;
+            statistics.NumberOfTimesPooled++;
+
+            // See if we're pooling something both large and sparse.
+            if (builder.Capacity > MinCapacityToConsiderThreshold &&
+                ((double)builder.Count / builder.Capacity) < SparseThresholdRatio)
+            {
+                // Console.WriteLine($"Pooled when sparse: {builder.GetType()} {builder.Count} / {builder.Capacity}");
+                statistics.NumberOfTimesPooledWhenSparse++;
+            }
+
+            builder.Clear();
+
+            // See if this builder has been consistently sparse. If so then time to lower its capacity.
+            if (statistics.NumberOfTimesPooled > MinTimesPooledToConsiderStatistics &&
+                ((double)statistics.NumberOfTimesPooledWhenSparse / statistics.NumberOfTimesPooled) > ConsistentlySparseRatio)
+            {
+                // Console.WriteLine($"Halved capacity: {builder.GetType()} {statistics.NumberOfTimesPooledWhenSparse} / {statistics.NumberOfTimesPooled}");
```

Comment:
Would it be worth having some trace events for tracking this?

Reply:
In general, it might be worth having some amount of etw logging to inform us if the numbers you've picked are optimal?

Reply:
do you have examples of where the compiler currently uses ETW so i can mimic what goes on there?

Reply:
done.  can you look if i did it right?